alerts/geomodel/locality.py

```diff
@@ -0,0 +1,255 @@
+    '''
+
+    _source = event.get('_source', {})
+
+    source_ip = _source.get('sourceipaddress')
+    geo_data = _source.get('sourceipgeolocation')
+
+    if source_ip is None or geo_data is None:
+        return None
+
+    now = datetime.strftime(datetime.utcnow(), '%Y-%m-%dT%H:%M:%S.%f+00:00')
+    active_time_str = _source.get('utctimestamp', now)
+    active_time = _parse_datetime(active_time_str)
+
+    (city, country, lat, lon) = (
+        geo_data.get('city'),
+        geo_data.get('country_code'),
+        geo_data.get('latitude'),
+        geo_data.get('longitude')
+    )
+
+    if any([v is None for v in [city, country, lat, lon]]):
+        return None
+
+    return Locality(source_ip, city, country, active_time, lat, lon, radius)
+
+
+def find(qes: QueryInterface, username: str, index: str) -> Optional[Entry]:
+    '''Retrieve the locality state for one user from ElasticSearch.
+    '''
+
+    search = SearchQuery()
+    search.add_must([
+        TermMatch('type_', 'locality'),
+        TermMatch('username', username)
+    ])
+
+    return qes(search, index)
+
+
+def update(state: State, from_evt: State) -> Update:
+    '''Update the localities stored under an existing `State` against those
+    contained in a new `State` constructed from events.
+    '''
+
+    did_update = False
+
+    for loc1 in from_evt.localities:
+        did_find = False
+
+        for index, loc2 in enumerate(state.localities):
+            # If we find that the new state's locality has been recorded
+            # for the user in question, we only want to update it if either
+            # their IP changed or the new time of activity is more recent.
+            if loc1.city == loc2.city and loc1.country == loc2.country:
+                did_find = True
+
+                new_more_recent = loc1.lastaction > loc2.lastaction
+                new_ip = loc1.sourceipaddress != loc2.sourceipaddress
+
+                if new_more_recent or new_ip:
+                    state.localities[index] = loc1
+                    did_update = True
+
+                # Stop looking for the locality in the records pulled from ES.
+                break
+
+        if not did_find:
+            state.localities.append(loc1)
+            did_update = True
+
+    return Update(state, did_update)
+
+
+def remove_outdated(state: State, days_valid: int) -> Update:
+    '''Update a state by removing localities that are outdated, determined
+    by checking if the last activity within a given locality was at least
+    some number of days ago.
+    '''
+
+    now = datetime.utcnow().replace(tzinfo=pytz.UTC)
+    last_valid_date = now - timedelta(days=days_valid)
+
+    new_localities = [
+        loc
+        for loc in state.localities
+        if loc.lastaction >= last_valid_date
+    ]
+
+    return Update(
+        state=State(state.type_, state.username, new_localities),
+        did_update=len(new_localities) != len(state.localities))
+
+
+def _parse_datetime(datetime_str):
```

Comment:
We shouldn't need to specify this logic, the event timestamps should be all in a guessable format, so we can simply call `toUTC` from https://github.com/mozilla/MozDef/blob/master/mozdef_util/mozdef_util/utilities/toUTC.py#L10